[java] StackOverflowError with recursively bound type variable

<!-- Please, prefix the report title with the language it applies to within brackets, such as [java] or [apex].
If not specific to a language, you can use [core]. -->

**Affects PMD Version:** 7.0.0, 7.1.0, 7.2.0, 7.3.0

**Description:**

I'm trying to upgrade PMD from 6.55.0 to 7.x but the PMD runs just end up in `java.lang.StackoverflowErrror`s.

`ruleset.xml`

```xml
<?xml version="1.0" encoding="UTF-8"?>
<ruleset name="PMD ruleset"
         xmlns="http://pmd.sourceforge.net/ruleset/2.0.0"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://pmd.sourceforge.net/ruleset/2.0.0 https://pmd.sourceforge.io/ruleset_2_0_0.xsd">

  <description>PMD ruleset</description>

  <rule ref="category/java/bestpractices.xml/MissingOverride"/>
  <rule ref="category/java/bestpractices.xml/PrimitiveWrapperInstantiation"/>
  <rule ref="category/java/bestpractices.xml/ReplaceVectorWithList"/>
  <rule ref="category/java/bestpractices.xml/UseStandardCharsets"/>
  <rule ref="category/java/codestyle.xml/ExtendsObject"/>
  <rule ref="category/java/codestyle.xml/PackageCase"/>
  <rule ref="category/java/codestyle.xml/UnnecessaryCast"/>
  <rule ref="category/java/codestyle.xml/UnnecessarySemicolon"/>
</ruleset>
```
But also fails with just one of the rules in place. 

Gradle configuration:
```groovy
apply plugin: 'pmd'

pmd {
    toolVersion = '7.3.0'
    consoleOutput = true
    incrementalAnalysis = true
    ruleSets = [ "$rootProject.projectDir/gradle/pmd/ruleset.xml" ]
}
```

`gradle.properties`:
```ini
org.gradle.parallel=true
org.gradle.jvmargs=-Xmx3g
```

Tested with both Temurin 17 and 21 (the project requires at least 17) on Ubuntu 22.04.

**Exception Stacktrace:**

The topmost part of the stack trace inline, [here is the full stack trace](https://github.com/user-attachments/files/16068013/stacktrace.txt).

```
Caused by: java.lang.StackOverflowError
        at net.sourceforge.pmd.lang.java.types.Substitution.apply(Substitution.java:45)
        at net.sourceforge.pmd.lang.java.types.Substitution.apply(Substitution.java:28)
        at net.sourceforge.pmd.lang.java.types.SubstVar.subst(SubstVar.java:25)
        at net.sourceforge.pmd.lang.java.types.WildcardTypeImpl.subst(WildcardTypeImpl.java:35)
        at net.sourceforge.pmd.lang.java.types.WildcardTypeImpl.subst(WildcardTypeImpl.java:17)
        at net.sourceforge.pmd.lang.java.types.TypeOps.lambda$subst$0(TypeOps.java:950)
        at net.sourceforge.pmd.lang.java.types.TypeOps.mapPreservingSelf(TypeOps.java:977)
        at net.sourceforge.pmd.lang.java.types.TypeOps.subst(TypeOps.java:950)
        at net.sourceforge.pmd.lang.java.types.JClassType.subst(JClassType.java:78)
        at net.sourceforge.pmd.lang.java.types.JClassType.subst(JClassType.java:54)
        at net.sourceforge.pmd.lang.java.types.TypeOps.subst(TypeOps.java:941)
        at net.sourceforge.pmd.lang.java.types.TypeConversion.capture(TypeConversion.java:213)
        at net.sourceforge.pmd.lang.java.types.TypeConversion.capture(TypeConversion.java:154)
        at net.sourceforge.pmd.lang.java.types.TypeConversion.capture(TypeConversion.java:152)
        at net.sourceforge.pmd.lang.java.types.TypeOps$SubtypeVisitor.isConvertible(TypeOps.java:662)
        at net.sourceforge.pmd.lang.java.types.TypeOps$SubtypeVisitor.isConvertible(TypeOps.java:598)
        at net.sourceforge.pmd.lang.java.types.TypeOps.isConvertiblePure(TypeOps.java:415)
        at net.sourceforge.pmd.lang.java.types.TypeOps.mostSpecific(TypeOps.java:1757)
        at net.sourceforge.pmd.lang.java.types.Lub.glb(Lub.java:316)
        at net.sourceforge.pmd.lang.java.types.TypeSystem.glb(TypeSystem.java:723)
        at net.sourceforge.pmd.lang.java.types.TypeConversion.capture(TypeConversion.java:227)
        at net.sourceforge.pmd.lang.java.types.TypeConversion.capture(TypeConversion.java:154)
        at net.sourceforge.pmd.lang.java.types.TypeConversion.capture(TypeConversion.java:152)
        at net.sourceforge.pmd.lang.java.types.TypeOps$SubtypeVisitor.isConvertible(TypeOps.java:662)
        at net.sourceforge.pmd.lang.java.types.TypeOps$SubtypeVisitor.isConvertible(TypeOps.java:598)
        at net.sourceforge.pmd.lang.java.types.TypeOps.isConvertiblePure(TypeOps.java:415)
        at net.sourceforge.pmd.lang.java.types.TypeOps.mostSpecific(TypeOps.java:1757)
        at net.sourceforge.pmd.lang.java.types.Lub.glb(Lub.java:316)
        at net.sourceforge.pmd.lang.java.types.TypeSystem.glb(TypeSystem.java:723)
        at net.sourceforge.pmd.lang.java.types.TypeConversion.capture(TypeConversion.java:227)
        at net.sourceforge.pmd.lang.java.types.TypeConversion.capture(TypeConversion.java:154)
        at net.sourceforge.pmd.lang.java.types.TypeConversion.capture(TypeConversion.java:152)
        at net.sourceforge.pmd.lang.java.types.TypeOps$SubtypeVisitor.isConvertible(TypeOps.java:662)
        at net.sourceforge.pmd.lang.java.types.TypeOps$SubtypeVisitor.isConvertible(TypeOps.java:598)
        at net.sourceforge.pmd.lang.java.types.TypeOps.isConvertiblePure(TypeOps.java:415)
        at net.sourceforge.pmd.lang.java.types.TypeOps.mostSpecific(TypeOps.java:1757)
        at net.sourceforge.pmd.lang.java.types.Lub.glb(Lub.java:316)
        at net.sourceforge.pmd.lang.java.types.TypeSystem.glb(TypeSystem.java:723)
```

**Code Sample demonstrating the issue:**

I cannot share the project as is, and I haven't yet been able to figure out what exactly triggers it. Any help in pinpointing the problematic class would be appreciated.

**Steps to reproduce:**

TBD.

**Running PMD through:** *Gradle 8.8*

--- a/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/types/JTypeVar.java
+++ b/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/types/JTypeVar.java
@@ -14,6 +14,7 @@ import org.checkerframework.checker.nullness.qual.NonNull;
 import org.checkerframework.checker.nullness.qual.Nullable;
 import org.pcollections.PSet;
 
+import net.sourceforge.pmd.annotation.InternalApi;
 import net.sourceforge.pmd.lang.java.symbols.JMethodSymbol;
 import net.sourceforge.pmd.lang.java.symbols.JTypeParameterSymbol;
 import net.sourceforge.pmd.lang.java.symbols.SymbolicValue.SymAnnot;
@@ -112,17 +113,26 @@ public interface JTypeVar extends SubstVar {
 
     /**
      * @throws UnsupportedOperationException If this is not a capture var
+     *
+     * <p>Note that this is only supposed to be used internally.
      */
+    @InternalApi
     JTypeVar cloneWithBounds(JTypeMirror lower, JTypeMirror upper);
 
     /**
      * Return a new type variable with the same underlying symbol or
      * capture variable, but the upper bound is now the given type.
      *
+     * <p>Note that this is only supposed to be used internally. For
+     * now it only serves to apply type annotations to the upper bound
+     * when parsing class files. Some implementations may therefore throw
+     * {@link UnsupportedOperationException}.
+     *
      * @param newUB New upper bound
      *
      * @return a new tvar
      */
+    @InternalApi
     JTypeVar withUpperBound(@NonNull JTypeMirror newUB);
 
     @Override // refine return type

--- a/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/types/JTypeVar.java
+++ b/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/types/JTypeVar.java
@@ -14,7 +14,6 @@ import org.checkerframework.checker.nullness.qual.NonNull;
 import org.checkerframework.checker.nullness.qual.Nullable;
 import org.pcollections.PSet;
 
-import net.sourceforge.pmd.annotation.InternalApi;
 import net.sourceforge.pmd.lang.java.symbols.JMethodSymbol;
 import net.sourceforge.pmd.lang.java.symbols.JTypeParameterSymbol;
 import net.sourceforge.pmd.lang.java.symbols.SymbolicValue.SymAnnot;
@@ -116,7 +115,6 @@ public interface JTypeVar extends SubstVar {
      *
      * <p>Note that this is only supposed to be used internally.
      */
-    @InternalApi
     JTypeVar cloneWithBounds(JTypeMirror lower, JTypeMirror upper);
 
     /**
@@ -132,7 +130,6 @@ public interface JTypeVar extends SubstVar {
      *
      * @return a new tvar
      */
-    @InternalApi
     JTypeVar withUpperBound(@NonNull JTypeMirror newUB);
 
     @Override // refine return type

--- a/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/types/TypeConversion.java
+++ b/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/types/TypeConversion.java
@@ -17,6 +17,7 @@ import java.util.List;
 
 import org.checkerframework.checker.nullness.qual.Nullable;
 
+import net.sourceforge.pmd.lang.java.types.TypeVarImpl.CapturedTypeVar;
 import net.sourceforge.pmd.lang.java.types.internal.infer.InferenceVar;
 import net.sourceforge.pmd.util.CollectionUtil;
 
@@ -145,11 +146,7 @@ public final class TypeConversion {
      */
     public static JTypeMirror capture(JTypeMirror t) {
         if (t instanceof JTypeVar) {
-            // Need to capture the upper bound because that bound contributes the methods of the tvar.
-            // Eg in `<C extends Collection<? super X>>` the methods available in C may mention the type
-            // parameter of collection. But this is a wildcard `? super X` that needs to be captured.
-            JTypeVar tv = (JTypeVar) t;
-            return tv.withUpperBound(capture(tv.getUpperBound()));
+            return TypeVarImpl.tvarCapture((JTypeVar) t);
         }
         return t instanceof JClassType ? capture((JClassType) t) : t;
     }
@@ -207,7 +204,7 @@ public final class TypeConversion {
 
             if (arg instanceof JWildcardType) {
                 JWildcardType w = (JWildcardType) arg;        // Ti alias
-                TypeVarImpl.CapturedTypeVar freshVar = (TypeVarImpl.CapturedTypeVar) fresh; // Si alias
+                CapturedTypeVar freshVar = (CapturedTypeVar) fresh; // Si alias
 
                 JTypeMirror prevUpper = wellFormed ? typeParams.get(i).getUpperBound() : ts.OBJECT; // Ui
                 JTypeMirror substituted = TypeOps.subst(prevUpper, subst);

--- a/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/types/TypeOps.java
+++ b/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/types/TypeOps.java
@@ -398,7 +398,7 @@ public final class TypeOps {
 
 
     public static Convertibility isConvertible(@NonNull JTypeMirror t, @NonNull JTypeMirror s) {
-        return SubtypeVisitor.INFERENCE.isConvertible(t, s);
+        return SubtypeVisitor.INFERENCE.isConvertible(t, s, true);
     }
 
     @Deprecated // unused
@@ -595,7 +595,7 @@ public final class TypeOps {
 
 
         Convertibility isConvertible(@NonNull JTypeMirror t, @NonNull JTypeMirror s) {
-            return isConvertible(t, s, true);
+            return isConvertible(t, s, false);
         }
 
         /**
@@ -1798,7 +1798,7 @@ public final class TypeOps {
         for (JTypeMirror v : set) {
             for (JTypeMirror w : set) {
                 if (!w.equals(v) && !hasUnresolvedSymbolOrArray(w)) {
-                    Convertibility isConvertible = isConvertiblePure(w, v);
+                    Convertibility isConvertible = isConvertibleNoCapture(w, v);
                     if (isConvertible.bySubtyping()
                         // This last case covers unchecked conversion. It is made antisymmetric by the
                         // test for a symbol. eg |G| <~> G<?> so it would fail.

--- a/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/types/TypeVarImpl.java
+++ b/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/types/TypeVarImpl.java
@@ -4,6 +4,8 @@
 
 package net.sourceforge.pmd.lang.java.types;
 
+import static net.sourceforge.pmd.lang.java.types.TypeConversion.capture;
+
 import java.util.Objects;
 import java.util.function.Function;
 
@@ -13,7 +15,6 @@ import org.pcollections.PSet;
 
 import net.sourceforge.pmd.lang.java.symbols.JTypeParameterSymbol;
 import net.sourceforge.pmd.lang.java.symbols.SymbolicValue.SymAnnot;
-import net.sourceforge.pmd.util.AssertionUtil;
 
 @SuppressWarnings("PMD.CompareObjectsWithEquals")
 abstract class TypeVarImpl implements JTypeVar {
@@ -59,8 +60,60 @@ abstract class TypeVarImpl implements JTypeVar {
      * the capture conversion algo in {@link TypeConversion#capture(JTypeMirror)}.
      * Captured variables use reference identity as equality relation.
      */
-    static TypeVarImpl.CapturedTypeVar freshCapture(JWildcardType wildcard) {
+    static TypeVarImpl.CapturedTypeVar freshCapture(@NonNull JWildcardType wildcard) {
         return new CapturedTypeVar(wildcard, wildcard.getTypeAnnotations());
+    }
+
+    /**
+     * Capture a type variable, that is, capture its bounds if needed.
+     * This is necessary because those bounds contributes to the methods
+     * of the type variable.
+     * Eg in {@code <C extends Collection<? super X>>} the methods available
+     * in C may mention the type parameter of Collection. But this is a
+     * wildcard `? super X` that needs to be replaced by a capture variable.
+     *
+     * @param tv a type var
+     */
+    static JTypeVar tvarCapture(@NonNull JTypeVar tv) {
+        if (tv.isCaptured()) {
+            return tv;
+        }
+        // Need to capture the bounds because those bounds contributes the methods of the tvar.
+        // Eg in `<C extends Collection<? super X>>` the methods available in C may mention the type
+        // parameter of collection. But this is a wildcard `? super X` that needs to be captured.
+        JTypeMirror upperBoundCap = capture(tv.getUpperBound());
+        JTypeMirror lowerBoundCap = capture(tv.getLowerBound());
+        if (upperBoundCap == tv.getUpperBound() && lowerBoundCap == tv.getLowerBound()) {
+            // no change
+            return tv;
+        }
+        // We will return a new var.
+        CapturedTypeVar newTv = new CapturedTypeVar(tv, upperBoundCap, lowerBoundCap, tv.getTypeAnnotations());
+
+        // We have to update the bounds again, to uphold recursive bounds. Eg if you have the following:
+        //    class C<T extends C<? extends T>>
+        // then capturing T in the body of the class C should produce a new capture var, call it T2,
+        // which has captured bounds. The upper bound of T2 should be
+        //  capture(C<? extends T2>)
+        // and notice that here it's T2 and not T, otherwise the recursive bound is
+        // not preserved by capture. So this last update is there to map T to T2 (ie, tv to newTv).
+        newTv.upperBound = upperBoundCap.subst(sv -> updateBounds(tv, sv, newTv));
+        newTv.lowerBound = lowerBoundCap.subst(sv -> updateBounds(tv, sv, newTv));
+        return newTv;
+    }
+
+    private static SubstVar updateBounds(JTypeVar tv, SubstVar sv, CapturedTypeVar newTv) {
+        if (sv == tv) {
+            return newTv;
+        } else if (sv instanceof JTypeVar) {
+            return ((JTypeVar) sv).substInBounds(sv2 -> {
+                if (sv2 == tv) {
+                    return newTv;
+                }
+                return sv2;
+            });
+        }
+        return sv;
     }
 
     static final class RegularTypeVar extends TypeVarImpl {
@@ -173,7 +226,7 @@ abstract class TypeVarImpl implements JTypeVar {
 
         @Override
         public int hashCode() {
-            return Objects.hash(symbol);
+            return symbol.hashCode();
         }
     }
 
@@ -181,19 +234,37 @@ abstract class TypeVarImpl implements JTypeVar {
 
         private static final int PRIME = 997;  // largest prime less than 1000
 
-        private final @NonNull JWildcardType wildcard;
+        private final @Nullable JWildcardType wildcard;
+        private final @Nullable JTypeVar tvar;
 
         private JTypeMirror upperBound;
         private JTypeMirror lowerBound;
 
-        private CapturedTypeVar(JWildcardType wild, PSet<SymAnnot> typeAnnots) {
+        private CapturedTypeVar(@NonNull JWildcardType wild, PSet<SymAnnot> typeAnnots) {
             this(wild, wild.asLowerBound(), wild.asUpperBound(), typeAnnots);
         }
 
-        private CapturedTypeVar(JWildcardType wild, @NonNull JTypeMirror lower, @NonNull JTypeMirror upper, PSet<SymAnnot> typeAnnots) {
-            super(wild.getTypeSystem(), typeAnnots);
+        private CapturedTypeVar(@Nullable JWildcardType wild, @NonNull JTypeMirror lower, @NonNull JTypeMirror upper, PSet<SymAnnot> typeAnnots) {
+            super(lower.getTypeSystem(), typeAnnots);
             this.upperBound = upper;
             this.lowerBound = lower;
+            this.wildcard = wild;
+            this.tvar = null;
+        }
+
+        private CapturedTypeVar(@Nullable JTypeVar tvar, @NonNull JTypeMirror lower, @NonNull JTypeMirror upper, PSet<SymAnnot> typeAnnots) {
+            super(lower.getTypeSystem(), typeAnnots);
+            this.upperBound = upper;
+            this.lowerBound = lower;
+            this.tvar = tvar;
+            this.wildcard = null;
+        }
+
+        private CapturedTypeVar(@Nullable JTypeVar tvar, JWildcardType wild, @NonNull JTypeMirror lower, @NonNull JTypeMirror upper, PSet<SymAnnot> typeAnnots) {
+            super(lower.getTypeSystem(), typeAnnots);
+            this.upperBound = upper;
+            this.lowerBound = lower;
+            this.tvar = tvar;
             this.wildcard = wild;
         }
 
@@ -223,7 +294,7 @@ abstract class TypeVarImpl implements JTypeVar {
 
         @Override
         public boolean isCaptureOf(JWildcardType wildcard) {
-            return this.wildcard.equals(wildcard);
+            return this.wildcard != null && this.wildcard.equals(wildcard);
         }
 
         @Override
@@ -233,11 +304,13 @@ abstract class TypeVarImpl implements JTypeVar {
 
         @Override
         public JTypeVar substInBounds(Function<? super SubstVar, ? extends @NonNull JTypeMirror> substitution) {
-            JWildcardType wild = this.wildcard.subst(substitution);
+            JWildcardType wild = this.wildcard == null ? null : this.wildcard.subst(substitution);
             JTypeMirror lower = getLowerBound().subst(substitution);
             JTypeMirror upper = getUpperBound().subst(substitution);
             if (wild == this.wildcard && lower == this.lowerBound && upper == this.lowerBound) {
                 return this;
+            } else if (wild == null) {
+                return new CapturedTypeVar(tvar, lower, upper, typeAnnots);
             }
             return new CapturedTypeVar(wild, lower, upper, getTypeAnnotations());
         }
@@ -253,13 +326,12 @@ abstract class TypeVarImpl implements JTypeVar {
             if (newTypeAnnots.isEmpty() && typeAnnots.isEmpty()) {
                 return this;
             }
-            return new CapturedTypeVar(wildcard, lowerBound, upperBound, newTypeAnnots);
+            return new CapturedTypeVar(tvar, wildcard, lowerBound, upperBound, newTypeAnnots);
         }
 
         @Override
         public JTypeVar withUpperBound(@NonNull JTypeMirror newUB) {
-            AssertionUtil.requireParamNotNull("upper bound", newUB);
-            return new CapturedTypeVar(wildcard, lowerBound, newUB, getTypeAnnotations());
+            throw new UnsupportedOperationException("This only needs to be implemented on regular type variables");
         }
 
         @Override
@@ -274,12 +346,13 @@ abstract class TypeVarImpl implements JTypeVar {
 
         @Override
         public @Nullable JTypeParameterSymbol getSymbol() {
-            return null;
+            return tvar == null ? null : tvar.getSymbol();
         }
 
         @Override
         public @NonNull String getName() {
-            return "capture#" + hashCode() % PRIME + " of " + wildcard;
+            Object captureOrigin = wildcard == null ? tvar : wildcard;
+            return "capture#" + hashCode() % PRIME + " of " + captureOrigin;
         }
     }
 }

--- a/pmd-java/src/test/kotlin/net/sourceforge/pmd/lang/java/types/AstTestUtil.kt
+++ b/pmd-java/src/test/kotlin/net/sourceforge/pmd/lang/java/types/AstTestUtil.kt
@@ -27,6 +27,6 @@ fun JavaNode.ctorCalls(): DescendantNodeStream<ASTConstructorCall> = descendants
 fun JavaNode.firstCtorCall() = ctorCalls().crossFindBoundaries().firstOrThrow()
 
 fun JavaNode.typeVariables(): MutableList<JTypeVar> = descendants(ASTTypeParameter::class.java).crossFindBoundaries().toList { it.typeMirror }
-fun JavaNode.varAccesses(name: String): NodeStream<ASTVariableAccess> = descendants(ASTVariableAccess::class.java).filter { it.name == name }
+fun JavaNode.varAccesses(name: String): NodeStream<ASTVariableAccess> = descendants(ASTVariableAccess::class.java).crossFindBoundaries().filter { it.name == name }
 fun JavaNode.varId(name: String) = descendants(ASTVariableId::class.java).filter { it.name == name }.firstOrThrow()
-fun JavaNode.typeVar(name: String) = descendants(ASTTypeParameter::class.java).filter { it.name == name }.firstOrThrow().typeMirror
+fun JavaNode.typeVar(name: String) = descendants(ASTTypeParameter::class.java).crossFindBoundaries().filter { it.name == name }.firstOrThrow().typeMirror

--- a/pmd-java/src/test/kotlin/net/sourceforge/pmd/lang/java/types/CaptureTest.kt
+++ b/pmd-java/src/test/kotlin/net/sourceforge/pmd/lang/java/types/CaptureTest.kt
@@ -7,9 +7,12 @@ package net.sourceforge.pmd.lang.java.types
 
 import io.kotest.core.spec.style.FunSpec
 import io.kotest.matchers.shouldBe
+import io.kotest.matchers.types.shouldBeSameInstanceAs
+import net.sourceforge.pmd.lang.java.ast.ASTVariableId
 import net.sourceforge.pmd.lang.java.symbols.internal.asm.createUnresolvedAsmSymbol
-import net.sourceforge.pmd.lang.java.types.TypeConversion.*
+import net.sourceforge.pmd.lang.java.types.TypeConversion.capture
 import net.sourceforge.pmd.lang.test.ast.IntelliMarker
+import net.sourceforge.pmd.lang.test.ast.shouldBeA
 
 /**
  * @author Clément Fournier
@@ -87,6 +90,39 @@ class CaptureTest : IntelliMarker, FunSpec({
 
                  capture(type) shouldBe box[captureMatcher(`?` extends child)]
             }
+
+            test("Capture of recursive types #5096") {
+                val acu = javaParser.parse(
+                    """
+
+                    class Child<C extends Child<? extends C>> {
+                      Child(C t) {
+                        super(t); // <-- The bug is triggered by capture of the `t` here
+                      } 
+                    }
+                """.trimIndent()
+                )
+
+                val (child) = acu.declaredTypeSignatures()
+
+                val parm = acu.descendants(ASTVariableId::class.java).firstOrThrow()
+
+                parm shouldHaveType child.typeArgs[0]
+                val captured = capture(parm.typeMirror)
+                captured.shouldBeA<JTypeVar> {
+                    it.isCaptured shouldBe true // its bounds are captured
+
+                    it.upperBound.shouldBeA<JClassType> {
+                        it.symbol shouldBe child.symbol
+                        it.typeArgs[0].shouldBeA<JTypeVar> { cvar ->
+                            cvar.isCaptured shouldBe true
+                            cvar.upperBound.shouldBeSameInstanceAs(captured)
+                        }
+                    }
+                }
+            }
+
+
 
         }
     }

--- a/pmd-java/src/test/kotlin/net/sourceforge/pmd/lang/java/types/GlbTest.kt
+++ b/pmd-java/src/test/kotlin/net/sourceforge/pmd/lang/java/types/GlbTest.kt
@@ -186,6 +186,29 @@ class GlbTest : FunSpec({
                     it.components.shouldContainExactly(tA, tB)
                 }
             }
+
+            test("Test GLB of recursive types #5096") {
+                val (acu, _) = javaParser.parseWithTypeInferenceSpy(
+                    """
+                        abstract class Settings<Q extends Settings<? extends Q, B>, B extends Settings.Builder<? extends B, Q>> implements Serializable {
+
+                            abstract static class Builder<B extends Builder<? extends B, S>, S extends Settings<? extends S, B>> extends DomainObjectBuilder<S, Modification<S>> {
+
+                                protected Builder(S bound, boolean boundToInstance) {
+                                    super(bound, boundToInstance);
+                                }
+                            }
+
+                            static class Modification<K extends Settings<? extends K, ?>> implements DomainObjectModification<K> {
+                            }
+                        }
+                    """.trimIndent()
+                )
+
+                acu.varAccesses("bound").firstOrThrow().typeMirror.shouldBeA<JTypeVar> {
+                    it.symbol shouldBe acu.typeVar("S").symbol
+                }
+            }
         }
     }
 

